momepy/preprocessing.py

```diff
@@ -740,3 +748,240 @@ def _get_extrapolated_line(coords, tolerance, point=False):
+           
+        rab_plus = rab_adj.drop(adj_long_polys)
+        
+    else:
+        rab['index_right'] = rab.index
+        rab_plus = rab
+    
+    rab_plus = rab_plus[['geometry', 'index_right']] #only keeping relevant fields
+    
+    return rab_plus
+
+def _rabs_center_points(gdf, center_type = 'centroid'):
+    """
+    From a selection of round abouts, returns an aggregated GeoDataFrame 
+    per round about with extra column with center_type. 
+    """
+    #creating a multipolygon per RAB (as opposed to dissolving) of the entire composition of the RAB
+    # temporary DataFrame where geometries is the array of pygeos geometries
+    # for aggregation to work properly
+    tmp = pd.DataFrame(gdf.copy())
+    tmp['geometry'] = tmp.geometry.values.data
+
+    rab_plus = gpd.GeoDataFrame(
+            tmp.groupby('index_right').geometry.apply(pygeos.multipolygons).rename('geometry'), crs= gdf.crs)
+    # verifying that all geometries are valid
+    rab_plus['geometry'] = rab_plus.apply(lambda row : make_valid(row.geometry) if not row.geometry.is_valid else row.geometry, axis = 1)
+    
+    if center_type == 'centroid' :
+        #geometry centroid of the actual circle
+        rab_plus['center_pt'] = gdf[gdf.index == gdf.index_right].geometry.centroid
+    
+    elif center_type == 'mean':
+        # mean geometry
+        ls_xy = [g.exterior.coords.xy for g in rab_plus.geometry] #extracting the points
+        
+        ls_xy = [point for polygon in multipolygon for point in polygon.exterior.coords[:-1]]
+        
+        mean_pts = [geometry.Point(np.mean(xy[0]),np.mean(xy[1])) for xy in ls_xy]
+        rab_plus['center_pt'] = gpd.GeoSeries( data = mean_pts, 
+                                              index = rab_plus.index, 
+                                              crs = edges.crs)
+    
+    # centerpoint of minimum_bounding_circle
+    # minimun_bounding_circle() not available in Shapely 1.8.2 but only in 'latest'
+    # --> https://shapely.readthedocs.io/en/latest/constructive.html
+    # current environment has 1.8.2
+    
+    return rab_plus
+
+def _coins_filtering_many_incoming(incoming_many, angle_threshold=0):
+    """
+    Used only for the cases when more than one incoming line touches the
+    roundabout.
+    """
+    # figuring out which one needs to be extended and retain attributes
+    coins_filter_result = []
+    # For each new connection, evaluate COINS and selecet the group from which the new line belongs
+    for g, x in incoming_many.groupby('line_wkt'):
+        gs = gpd.GeoSeries( pd.concat([x.geometry, x.line]), crs= incoming_many.crs )
+        gdf = gpd.GeoDataFrame(geometry = gs)
+        gdf.drop_duplicates(inplace=True)
+
+        coins = COINS(gdf, angle_threshold=angle_threshold)
+        stroke_gdf = coins.stroke_gdf()
+
+        orig_geom_join = stroke_gdf.sjoin(gpd.GeoDataFrame(geometry = x.line), predicate= 'covers' )
```

Comment:
Can we instead use `coins.stroke_attribute`? We could avoid the sjoin.

Reply:
The latest update now includes `coins.stroke_attribute` following the correction done last week by @amorfinv ! 
Much cleaner, indeed!